### ISSUE TITLE
False positive in "variable is never read" when using object.defineProperty

### ISSUE THREAD

**Issue by mastrzyz** [microsoft/TypeScript#40440]
I was originally looking at the migration guide on how to fix accessors overriding properties in TS -> https://github.com/microsoft/TypeScript/pull/33509#

Part of that guidance was using object.defineProperty and defining the getter/setter over there but hit a strange compiler error.

**TypeScript Version:**  3.8.3

**Search Terms:**
 - object define property false positive
- unread variable define property
**Code**

```ts
export class Foo {
  private _value: string | undefined;

  constructor() {
    Object.defineProperty(this, "value", {
      get() {
        return this._value;
      }
    });
  }
}
```

**Expected behavior:**

No compiler errors as I am defining a setter for the _value so the value is being read somewhere

**Actual behavior:**

'_value' is declared but its value is never read.

If I remove `_value` I get 

Property '_value' does not exist on type 'Foo'. Did you mean 'value'?

**Playground Link:** 
https://www.typescriptlang.org/play?ts=3.8.3#code/KYDwDg9gTgLgBAYwDYEMDOa4DEITgbwFgAoOOMKASwDcUZg4B9WpAV2AC440YqA7AOZwAPnFZ8AJsABmlPsAkBuEiTIIIfHlFYIY0ABQBKAqrJwA8gCMAVsF0A6KbPkAFKBDDBYAT30wAFpRoADRwAEQs7GGhRKRmZALAMEYmcfFkUEmsUHxwAUH2zChswMppZgC+pmQVhmU1JBVAA

**Related Issues:**

**Comment by j-oliveras** [comment]
Typescript don't understand that `this` in the getter definition will be the class. As a workaround, define the type of this inside the getter:
```ts
export class Foo {
  private _value: string | undefined;

  constructor() {
    Object.defineProperty(this, "value", {
      get(this: Foo) {
        return this._value;
      }
    });
  }
}
```

[Playground](https://www.typescriptlang.org/play?noUnusedLocals=true&ts=3.8.3#code/KYDwDg9gTgLgBAYwDYEMDOa4DEITgbwFgAoOOMKASwDcUZg4B9WpAV2AC440YqA7AOZwAPnFZ8AJsABmlPsAkBuEiTIIIfHlFYIY0ABQBKAqrJwA8gCMAVsF0A6KbPkAFKBDDBYAT30wAFpRoADRwAEQs7GGhRKRmZALAMH6BaFw4EMax8fFQSaxQfHABQfbMKGzAynHxAL6mZLWG1Y0ktUA)

**Comment by RyanCavanaugh** [comment]
👆
